### PR TITLE
Feature - TriNguyen - Navigation

### PR DESCRIPTION
**Pull request by mvn-tringuyen-dn**

### CODE REVIEW

**Comment by mvn-tanhuynh-dn** [comment]
Screen Shot 2022-05-06 at 9 04 09 AM
Can not build app, please check care fully before assign to review!

File: 09_Navigation/TriNC/Navigation/Navigation/AppDelegate/AppDelegate.swift [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+//        let appearance = UINavigationBarAppearance()
+//        appearance.configureWithOpaqueBackground()
+//        appearance.titleTextAttributes = [.foregroundColor: UIColor.white]
+//        appearance.backgroundColor = .red
+//
+//        UINavigationBar.appearance().standardAppearance = appearance
+//        UINavigationBar.appearance().scrollEdgeAppearance = appearance
+        return true
+    }
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/AppDelegate/AppDelegate.swift:22]: Remove un-use code

File: 09_Navigation/TriNC/Navigation/Navigation/EX1/EXA/ExerciseAViewController.swift [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+final class ExerciseAViewController: UIViewController {
+
+    // MARK: - Life cycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        title = "ViewController A"
+    }
+    
+    // MARK: - IBAction
+    @IBAction private func nextTouchUpInside(_ sender: Any) {
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX1/EXA/ExerciseAViewController.swift:19]: Function name must have full meaning!

File: 09_Navigation/TriNC/Navigation/Navigation/EX1/EXB/ExBViewController.swift [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    
+    // MARK: - IBActions
+    @IBAction private func nextButtonTouchUpInside(_ sender: Any) {
+        let vc = ExCViewController()
+        navigationController?.pushViewController(vc, animated: true)
+    }
+    
+    @IBAction private func previousButtonTouchUpInside(_ sender: Any) {
+        navigationController?.popViewController(animated: true)
+    }
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX1/EXB/ExBViewController.swift:26]: Function name must have full meaning!

File: 09_Navigation/TriNC/Navigation/Navigation/EX1/EXD/ExDViewController.swift [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    
+    // MARK: - IBActions
+    @IBAction private func nextButtonTouchUpinside(_ sender: Any) {
+        let vc = ExEViewController()
+        navigationController?.pushViewController(vc, animated: true)
+    }
+    
+    @IBAction private func cButtonTouchUpInside(_ sender: Any) {
+        guard let vc = (self.navigationController?.viewControllers[2]) else { return }
+        navigationController?.popToViewController(vc, animated: true)
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX1/EXD/ExDViewController.swift:26]: Don't use viewControllers[2],
Replace it by for loop and check controller isKind(of: ClassName.self)
(https://developer.apple.com/documentation/objectivec/nsobjectprotocol/1418511-iskind)

File: 09_Navigation/TriNC/Navigation/Navigation/EX1/EXD/ExDViewController.swift [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    
+    @IBAction private func bButtonTouchUpInside(_ sender: Any) {
+        guard let vc = (self.navigationController?.viewControllers[1]) else { return }
+        navigationController?.popToViewController(vc, animated: true)
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX1/EXD/ExDViewController.swift:31]: Same as above, replace it by for loop and check controller isKind(of: ClassName.self)

File: 09_Navigation/TriNC/Navigation/Navigation/EX1/EXE/ExEViewController.swift [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    // MARK: - Life cycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        title = "ViewController E"
+    }
+    
+    // MARK: - IBActions
+    @IBAction private func cButtonTouchUpInside(_ sender: Any) {
+        guard let vc = (self.navigationController?.viewControllers[2]) else { return }
+        navigationController?.popToViewController(vc, animated: true)
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX1/EXE/ExEViewController.swift:21]: Same as above, replace it by for loop and check controller isKind(of: ClassName.self)

File: 09_Navigation/TriNC/Navigation/Navigation/EX1/EXE/ExEViewController.swift [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    }
+    
+    @IBAction private func bButtonTouchUpInside(_ sender: Any) {
+        guard let vc = (self.navigationController?.viewControllers[1]) else { return }
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX1/EXE/ExEViewController.swift:25]: Same as above, replace it by for loop and check controller isKind(of: ClassName.self)

File: 09_Navigation/TriNC/Navigation/Navigation/EX2/DataManager/DataManager.swift [hunk 8]
```diff
@@ -8,12 +8,13 @@
 import Foundation
 
 final class DataManager {
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX2/DataManager/DataManager.swift:10]: Add space below class name

File: 09_Navigation/TriNC/Navigation/Navigation/EX2/Edit/EditViewController.swift [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+//  EditViewController.swift
+//  Navigation
+//
+//  Created by tri.nguyen on 19/04/2022.
+//
+
+import UIKit
+
+// MARK: - Protocol
+protocol EditViewControllerDelegate: class {
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX2/Edit/EditViewController.swift:11]: Add space below protocol name

File: 09_Navigation/TriNC/Navigation/Navigation/EX2/Edit/EditViewController.swift [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    // MARK: - Objc
+    @objc private func cancelAction() {
+        navigationController?.popViewController(animated: true)
+    }
+    
+    @objc private func doneAction() {
+        guard let username = usernameTextField.text else { return }
+        if let delegate = delegate {
+            delegate.didTapEditViewController(view: self, needsPeform: .tapDone(username: username))
+        }
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX2/Edit/EditViewController.swift:52]: Don't need unwrap delegate

File: 09_Navigation/TriNC/Navigation/Navigation/EX2/Edit/EditViewController.swift [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        navigationController?.popViewController(animated: true)
+    }
+}
+
+// MARK: - Extention
+extension EditViewController {
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX2/Edit/EditViewController.swift:58]: Add space below

File: 09_Navigation/TriNC/Navigation/Navigation/EX2/Home/HomeViewController.swift [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+//  HomeViewController.swift
+//  Navigation
+//
+//  Created by tri.nguyen on 19/04/2022.
+//
+
+import UIKit
+
+// MARK: - Protocol
+protocol HomeViewControllerDelegate: class {
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX2/Home/HomeViewController.swift:11]: Add space below

File: 09_Navigation/TriNC/Navigation/Navigation/EX2/Home/HomeViewController.swift [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    func homeController(view: HomeViewController, needsPerform action: HomeViewController.Action)
+}
+
+final class HomeViewController: UIViewController {
+
+    // MARK: - IBOutlet
+    @IBOutlet private weak var welcomeLabel: UILabel!
+
+    // MARK: - Properties
+    var username: String = ""
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX2/Home/HomeViewController.swift:21]: Add private keyword

File: 09_Navigation/TriNC/Navigation/Navigation/EX2/Home/HomeViewController.swift [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    @objc private func editingAction() {
+        let vc = EditViewController()
+        vc.delegate = self
+        navigationController?.pushViewController(vc, animated: true)
+    }
+    
+    @objc private func logoutAction() {
+        if let delegate = delegate {
+            delegate.homeController(view: self, needsPerform: .logout)
+        }
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX2/Home/HomeViewController.swift:51]: Don't need unwrap delegate

File: 09_Navigation/TriNC/Navigation/Navigation/EX2/Home/HomeViewController.swift [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+// MARK: - Extention
+extension HomeViewController {
+    // MARK: - Define Action
+    enum Action {
+        case logout
+    }
+}
+
+// MARK: - Implement EditProtocol
+extension HomeViewController: EditViewControllerDelegate {
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX2/Home/HomeViewController.swift:65]: Add space below

File: 09_Navigation/TriNC/Navigation/Navigation/EX2/Home/HomeViewController.swift [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    @objc private func logoutAction() {
+        if let delegate = delegate {
+            delegate.homeController(view: self, needsPerform: .logout)
+        }
+        navigationController?.popViewController(animated: true)
+    }
+}
+
+// MARK: - Extention
+extension HomeViewController {
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX2/Home/HomeViewController.swift:57]: Add space below

File: 09_Navigation/TriNC/Navigation/Navigation/EX3/Avatar/AvatarView.swift [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+//  Navigation
+//
+//  Created by tri.nguyen on 20/04/2022.
+//
+
+import Foundation
+import UIKit
+
+// MARK: - Protocol
+protocol AvatarViewDelegate: class {
```
Comment (mvn-tanhuynh-dn) [inline 09_Navigation/TriNC/Navigation/Navigation/EX3/Avatar/AvatarView.swift:12]: Add space below

File: 09_Navigation/TriNC/Navigation/Navigation/EX2/Login/LoginViewController.swift [hunk 18]
```diff
@@ -52,6 +49,12 @@ final class LoginViewController: UIViewController {
             notificationLabel.text = "Wrong username or password"
```
[the remark this answers is not in the document]
Comment (mvn-tringuyen-dn) [reply 09_Navigation/TriNC/Navigation/Navigation/EX2/Login/LoginViewController.swift:49]: The topic only asks to check if the textfield has data at the login screen, if there is data, then switch to the home screen.

**Review by mvn-tanhuynh-dn — APPROVED** [review]
LGTM!

**Pull request merged**